### ISSUE TITLE
ClassNLLCriterion loss gets more negative over training iterations

### ISSUE THREAD

**Issue by marcel1991** [torch/nn#1078]
When I train my model I plot the loss over the iterations it is doing. There I see the negative log loss to start at zero and then it becomes more negative. 

![lossplot](https://cloud.githubusercontent.com/assets/16177860/21452825/f5e852b4-c909-11e6-9b4e-d5c79c2f26b3.png)
As far as I understand it right from the documentation the loss is = -log(p_target) where p_target is the softmax probability of the output layer at the target. So if my output becomes better I would get a softmax probability approaching 1 at the target output or approaching 0 if my output is wrong. Therefore the negative log would be 0 or +INF respectively. However, I get a loss that starts at zero and gets more and more negative when the model is improving... So why is that? To my understanding it should start at +INF and then approach zero without every getting below zero.

I tested it on the validation set and it is indeed giving way better accuracies after some iterations so I guess it is still learning something.

The batch training function is below. The batchLoss in this case always gives negative values.
``` lua
function TrainManager.trainBatch(self, batchInputsCpu, batchLabelsCpu)
  cutorch.synchronize()
  collectgarbage() -- free unused memory

  local options = self.options
  local model = self.model
  local modelParameters = self.modelParameters
  local gradientParameters = self.gradientParameters
  local criterion = self.criterion
  local optimisationState = self.optimisationState
  local batchInputs = self.batchInputs
  local batchLabels = self.batchLabels

  -- copy data into gpu tensors
  batchInputs:resize(batchInputsCpu:size()):copy(batchInputsCpu)
  batchLabels:resize(batchLabelsCpu:size()):copy(batchLabelsCpu)

  local batchLoss
  -- sgd expects function with input: moduleParameters; output: loss, gradParams
  local opFunction = function(params)
    model:zeroGradParameters()

    local outputs = model:forward(batchInputs)
    batchLoss = criterion:forward(outputs, batchLabels)
    local gradientOutputs = criterion:backward(outputs, batchLabels)
    model:backward(batchInputs, gradientOutputs)

    return batchLoss, gradientParameters
  end

  optim.sgd(opFunction, modelParameters, optimisationState)
  cutorch.synchronize()

  print("Loss: " .. batchLoss)
  ...

end
```

My model is a VGG16 and looks as follows:

````lua
 (1): cudnn.SpatialConvolution(3 -> 64, 3x3, 1,1, 1,1)
  (2): cudnn.ReLU
  (3): cudnn.SpatialConvolution(64 -> 64, 3x3, 1,1, 1,1)
  (4): cudnn.ReLU
  (5): cudnn.SpatialMaxPooling(2x2, 2,2)
  (6): cudnn.SpatialConvolution(64 -> 128, 3x3, 1,1, 1,1)
  (7): cudnn.ReLU
  (8): cudnn.SpatialConvolution(128 -> 128, 3x3, 1,1, 1,1)
  (9): cudnn.ReLU
  (10): cudnn.SpatialMaxPooling(2x2, 2,2)
  (11): cudnn.SpatialConvolution(128 -> 256, 3x3, 1,1, 1,1)
  (12): cudnn.ReLU
  (13): cudnn.SpatialConvolution(256 -> 256, 3x3, 1,1, 1,1)
  (14): cudnn.ReLU
  (15): cudnn.SpatialConvolution(256 -> 256, 3x3, 1,1, 1,1)
  (16): cudnn.ReLU
  (17): cudnn.SpatialMaxPooling(2x2, 2,2)
  (18): cudnn.SpatialConvolution(256 -> 512, 3x3, 1,1, 1,1)
  (19): cudnn.ReLU
  (20): cudnn.SpatialConvolution(512 -> 512, 3x3, 1,1, 1,1)
  (21): cudnn.ReLU
  (22): cudnn.SpatialConvolution(512 -> 512, 3x3, 1,1, 1,1)
  (23): cudnn.ReLU
  (24): cudnn.SpatialMaxPooling(2x2, 2,2)
  (25): cudnn.SpatialConvolution(512 -> 512, 3x3, 1,1, 1,1)
  (26): cudnn.ReLU
  (27): cudnn.SpatialConvolution(512 -> 512, 3x3, 1,1, 1,1)
  (28): cudnn.ReLU
  (29): cudnn.SpatialConvolution(512 -> 512, 3x3, 1,1, 1,1)
  (30): cudnn.ReLU
  (31): cudnn.SpatialMaxPooling(2x2, 2,2)
  (32): nn.View(-1)
  (33): nn.Linear(25088 -> 4096)
  (34): cudnn.ReLU
  (35): nn.Dropout(0.500000)
  (36): nn.Linear(4096 -> 4096)
  (37): cudnn.ReLU
  (38): nn.Dropout(0.500000)
  (39): nn.Linear(4096 -> 1000)
  (40): cudnn.SoftMax
```

**Comment by eriche2016** [comment]
You should use nn.LogSoftMax instead of nn.SoftMax before using ClassNLL criterion.

**Comment by marcel1991** [comment]
Ah yeah that makes sense. That's why I'm starting at zero and why I'm approaching -1. I gonna try it out later and tell you if it worked.

**Comment by kmul00** [comment]
https://github.com/torch/nn/blob/master/doc/criterion.md#nn.ClassNLLCriterion

2nd para.

**Comment by marcel1991** [comment]
Thanks, that was the solution!
